Implement hit-testing changes for transform-style:preserve-3d and perspective to only apply to direct DOM children.

mattwoodrow opened the pull request:
#### dc9396c615610f61cdefb62e3a014ee111ef3a76
<pre>
Implement hit-testing changes for transform-style:preserve-3d and perspective to only apply to direct DOM children.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248283">https://bugs.webkit.org/show_bug.cgi?id=248283</a>
rdar://problem/102632592&gt;

Reviewed by Simon Fraser.

Unfortunately this is fairly complex, due to needing to support both the new and old behaviour and switch at runtime.

The new behaviour is that all hit-testing methods return both the hit layer, as well as the depth that it was hit at.
There&apos;s no shared mutable state between calls (zOffset, zOffsetForDescendantsPtr etc), and each function is separately
responsible for determing the frontmost layer if sorting is enabled for that context.

hitTestList detects when we&apos;re about to recurse into a child that doesn&apos;t share the same 3d coordinate space as the current
layer, and passes down a flattened transformState if so. It also then overrides the returned hit depth with that of the plane
that we flattened into.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/3d-rendering-context-behavior-expected.txt:
* LayoutTests/platform/mac/transforms/3d/point-mapping/3d-point-mapping-2-expected.txt:
* LayoutTests/platform/gtk/transforms/3d/point-mapping/3d-point-mapping-2-expected.txt:
* LayoutTests/platform/ios/transforms/3d/point-mapping/3d-point-mapping-2-expected.txt:

New test passing, but also two newly failing due to point-mapping and hit-testing being out of sync.
These will be fixed in a follow-up to implement the point-mapping changes.

* LayoutTests/transforms/3d/hit-testing/overlapping-layers-hit-test-expected.txt:
* LayoutTests/transforms/3d/hit-testing/overlapping-layers-hit-test.html:

New subtest added that has a empty spacer div inbetween the element with preserve-3d and the transformed
descendants, confirms that we hit test this without the transform being accumulated.

* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2-expected.txt: Added.
* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2.html: Added.

New test with perspective separated from the transformed descendant with a spacer.

* LayoutTests/transforms/3d/point-mapping/3d-point-mapping-2.html:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::hitTest):
(WebCore::parentLayerIsDOMParent):
(WebCore::isHitCandidateLegacy):
(WebCore::RenderLayer::participatesInPreserve3D const):
(WebCore::RenderLayer::hitTestLayer):
(WebCore::RenderLayer::hitTestTransformedLayerInFragments):
(WebCore::RenderLayer::hitTestLayerByApplyingTransform):
(WebCore::RenderLayer::hitTestList):
(WebCore::isHitCandidate): Deleted.
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldUseTransformFromContainer const):
(WebCore::RenderObject::getTransformFromContainer const):

Only include the perspective transform if the container with perspective is
the direct DOM parent.

Canonical link: <a href="https://commits.webkit.org/257255@main">https://commits.webkit.org/257255@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3fff444849ea56e81752d7238a3e6617de880505

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98191 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/7404 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31336 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107652 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167931 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102133 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7898 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/36170 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90780 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/104238 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103843 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5934 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84787 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33046 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87816 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89513 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/75983 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1368 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/20945 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1325 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22448 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4997 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6201 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44958 "Found 2 new test failures: http/wpt/service-workers/fetch-service-worker-preload-download.https.html, http/wpt/service-workers/fetch-service-worker-preload.https.html (failure)") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2662 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41872 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.h:
```diff
@@ -1107,13 +1109,14 @@ class RenderLayer : public CanMakeWeakPtr<RenderLayer> {
     void beginTransparencyLayers(GraphicsContext&, const LayerPaintingInfo&, const LayoutRect& dirtyRect);
 
-    RenderLayer* hitTestLayer(RenderLayer* rootLayer, RenderLayer* containerLayer, const HitTestRequest&, HitTestResult&,
+    using HitLayer = std::pair<RenderLayer*, double>;
```
Hard to know what the double represents here. A struct would be clearer.

smfr replied:
Also if you used a struct you could initialize the z param to 0.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3842,14 +3842,14 @@ bool RenderLayer::hitTest(const HitTestRequest& request, const HitTestLocation&
 
-    RenderLayer* insideLayer = hitTestLayer(this, nullptr, request, result, hitTestArea, hitTestLocation, false);
-    if (!insideLayer) {
+    HitLayer insideLayer = hitTestLayer(this, nullptr, request, result, hitTestArea, hitTestLocation, false);
```
auto

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4075,33 +4089,41 @@ RenderLayer* RenderLayer::hitTestLayer(RenderLayer* rootLayer, RenderLayer* cont
+
     // This variable tracks which layer the mouse ends up being inside.
-    RenderLayer* candidateLayer = nullptr;
+    HitLayer candidateLayer = { nullptr, -std::numeric_limits<double>::infinity() };
```
auto

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4214,25 +4245,24 @@ RenderLayer* RenderLayer::hitTestTransformedLayerInFragments(RenderLayer* rootLa
             continue;
 
-        RenderLayer* hitLayer = hitTestLayerByApplyingTransform(rootLayer, containerLayer, request, result, hitTestRect, hitTestLocation,
+        HitLayer hitLayer = hitTestLayerByApplyingTransform(rootLayer, containerLayer, request, result, hitTestRect, hitTestLocation,
```
auto

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4279,40 +4309,65 @@ bool RenderLayer::hitTestContents(const HitTestRequest& request, HitTestResult&
+        return { nullptr, 0 };
 
-    RenderLayer* resultLayer = nullptr;
+    HitLayer resultLayer = { nullptr, -std::numeric_limits<double>::infinity() };
```
auto resultLayer = HitLayer { ... }

mattwoodrow committed:
Implement hit-testing changes for transform-style:preserve-3d and perspective to only apply to direct DOM children.
https://bugs.webkit.org/show_bug.cgi?id=248283
rdar://problem/102632592>

Reviewed by Simon Fraser.

Unfortunately this is fairly complex, due to needing to support both the new and old behaviour and switch at runtime.

The new behaviour is that all hit-testing methods return both the hit layer, as well as the depth that it was hit at.
There's no shared mutable state between calls (zOffset, zOffsetForDescendantsPtr etc), and each function is separately
responsible for determing the frontmost layer if sorting is enabled for that context.

hitTestList detects when we're about to recurse into a child that doesn't share the same 3d coordinate space as the current
layer, and passes down a flattened transformState if so. It also then overrides the returned hit depth with that of the plane
that we flattened into.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/3d-rendering-context-behavior-expected.txt:
* LayoutTests/platform/mac/transforms/3d/point-mapping/3d-point-mapping-2-expected.txt:
* LayoutTests/platform/gtk/transforms/3d/point-mapping/3d-point-mapping-2-expected.txt:
* LayoutTests/platform/ios/transforms/3d/point-mapping/3d-point-mapping-2-expected.txt:

New test passing, but also two newly failing due to point-mapping and hit-testing being out of sync.
These will be fixed in a follow-up to implement the point-mapping changes.

* LayoutTests/transforms/3d/hit-testing/overlapping-layers-hit-test-expected.txt:
* LayoutTests/transforms/3d/hit-testing/overlapping-layers-hit-test.html:

New subtest added that has a empty spacer div inbetween the element with preserve-3d and the transformed
descendants, confirms that we hit test this without the transform being accumulated.

* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2-expected.txt: Added.
* LayoutTests/transforms/3d/hit-testing/rotated-hit-test-2.html: Added.

New test with perspective separated from the transformed descendant with a spacer.

* LayoutTests/transforms/3d/point-mapping/3d-point-mapping-2.html:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::hitTest):
(WebCore::parentLayerIsDOMParent):
(WebCore::isHitCandidateLegacy):
(WebCore::RenderLayer::participatesInPreserve3D const):
(WebCore::RenderLayer::hitTestLayer):
(WebCore::RenderLayer::hitTestTransformedLayerInFragments):
(WebCore::RenderLayer::hitTestLayerByApplyingTransform):
(WebCore::RenderLayer::hitTestList):
(WebCore::isHitCandidate): Deleted.
* Source/WebCore/rendering/RenderLayer.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::shouldUseTransformFromContainer const):
(WebCore::RenderObject::getTransformFromContainer const):

Only include the perspective transform if the container with perspective is
the direct DOM parent.

Canonical link: https://commits.webkit.org/257255@main
